uClinux-dist/user/tripwire/man/tw.config.5
.\" $Id: tw.config.5,v 1.2 2005/07/25 06:51:34 magicyang Exp $
.\"
.\" Macros added by addw@phcomp.co.uk for those systems where the an
.\" troff macro package doesn't know about .Ps & .Pe.
.\" Note it is assumed that CW is a constant width font.
.\" Ps - Start display text
.de Ps
.nf
.in +0.5i
.ft CW
..
.\" Pe - end of display text
.de Pe
.fi
.in -0.5i
.ft 1
..
.\"
.TH TW.CONFIG 5 "October 5, 1992"
.SH NAME
tw.config \- configuration file for Tripwire
.SH SYNOPSIS
tw.config
.SH DESCRIPTION
The \fItw.config\fP file contains the list of files and
directories to be scanned by Tripwire.  Information on these
files is collected and stored in the \fItw.db\fP database file.
Stored with each \fItw.config\fP entry is a \fIselection-mask\fP
that describes what changes Tripwire can safely ignore without
reporting to the user (e.g., access timestamp).
.LP
The first section in this manual page describes the entry format in
\fItw.config\fP for the files monitored by Tripwire.  The second
section describes the proprocessing directives that Tripwire
provides.  These directives, which provide functionality similar to
the C preprocessor and M4 macro processor, allow Tripwire to make
bindings at run-time.  This allows system administrators to use common
\fItw.config\fP files across multiple machines \- or even across an
entire site.
.SH ENTRY FORMAT
.LP
Each entry in \fItw.config\fP is a single line in the following form:
.sp
.RS
\fBFormat\fP: 	\fI[!|=] entry [select-flags | template] [# comment]\fP
.RE
.RS
.TP 12
.I entry
An \fIentry\fP is the absolute pathname of a file or a directory. 
Without any prefixes, the \fIentry\fP is added to the list of files
to be scanned.  
.TP
.I \ 
Note that directories listed in the \fItw.config\fP file are
recursively descended.  However, filesystems are never crossed.
(I.e., if \fI/usr\fP and \fI/usr/local\fP are seperate filesystems, a
\fI/usr\fP in \fItw.config\fP entry will not scan files that reside
in the \fI/usr/local\fP filesystem.)
.RS
.TP
.I !
Inclusive prune.  Prunes \fIentry\fP from the list of files to be
scanned.  If \fIentry\fP is a file, the file is removed from the list
of files.  If \fIentry\fP is a directory, the directory and all of its
children are removed from the list of files.
.TP
.I =
Exclusive prune.  Does not prune \fIentry\fP, but does prune its
children.  This has no effect if \fIentry\fP is a file.  This option
is useful for monitoring directories with transient files (e.g., /tmp
and /var/tmp).
.RE
.TP
.I select-flags
\fIselect-flags\fP describe inode and file attributes.
\fIselect-flags\fP either specifies Tripwire to ignore changes in a
specific attribute, or to report them.  \fIselect-flags\fP are
provided in the form:
.I
[ [+|-][pinugsam123456789] ... ]
.RS
.TP 4
.I -
ignore the following attributes 	
.TP
.I +
record and check the following attributes
.TP
.I p
permission and file mode bits 	
.TP
.I i
inode number			
.TP
.I n
number of links (i.e., inode reference count)	
.TP
.I u
user id of owner			
.TP
.I g
group id of owner			
.TP
.I s
size of file
.TP
.I a
access timestamp
.TP
.I m
modification timestamp
.TP
.I c
inode creation/modification timestamp
.TP
.I 0
signature 0 \- null signature
.TP
.I 1
signature 1 \- MD5, the RSA Data Security, Inc. Message Digesting
Algorithm.
.TP
.I 2
signature 2 \- Snefru, the Xerox Secure Hash Function.
.TP
.I 3
signature 3 \- CRC-32, POSIX 1003.2 compliant 32-bit Cyclic Redundancy
Check.
.TP
.I 4
signature 4 \- CRC-16, the standard (non-CCITT) 16-bit Cyclic
Redundancy Check.
.TP
.I 5
signature 5 \- MD4, the RSA Data Security, Inc. Message Digesting
Algorithm.
.TP
.I 6
signature 6 \- MD2, the RSA Data Security, Inc. Message Digesting
Algorithm.
.TP
.I 7
signature 7 \- SHA, the NIST Secure Hash Algorithm (NIST FIPS 180)
.TP
.I 8
signature 8 \- Haval, a strong 128-bit signature algorithm
.TP
.I 9
signature 9 \- null signature (reserved for future expansion)
.RE
.TP
.I templates
\fItemplate\fP are predefined sets of \fIselect-flags\fP that are
commonly used by system administrators.
.RE
.LP
The following templates have been pre-defined to make these long 
\fIselect-masks\fP descriptions unnecessary.
.RS
.nf
.TP 6
.I R 
[R]ead-only (+pinugsm12-ac3456789) \fI(default)\fP
.TP
.I L 
[L]og file (+pinug-sacm123456789)
.TP
.I N
ignore [N]othing (+pinusgsamc123456789)
.TP
.I E
ignore [E]verything (-pinusgsamc123456789)
.TP
.I >
monotonically growing file (+pinug>-samc1233456789) \- the ``>'' indicates
that file changes are ignored only when the file is smaller than the last
recorded size.  This is useful for log files that are expected to grow.
.fi
.RE
.LP
By default, Tripwire uses the R template.  Because it applies the
set of \fIselect-flags\fP {+pinugsm12-a3456789}, Tripwire ignores
those changed files where only the access timestamp changed.
.LP
You can combine the use of templates with \fIselect-flag\fP modifiers.
The following entry monitors only changes in user-id and group-id
information.
.RS
/etc/lp	E+ug
.RE
.SH ENTRY EXAMPLES
.LP
The following entry will scan all the files in /etc, and report
any changes in mode bits, inode number, reference count, uid,
gid, modification and creation timestamp, and the signatures.
However, it will ignore any changes in the access timestamp.
.RS
/etc	+pinugsm12-a
.RE
.LP
It is equivalent to:
.RS
/etc	R
.RE
.LP
The following example shows a very simple \fItw.config\fP file that 
monitors selected directories.
.RS
.nf
/etc		R	# all system files
!/etc/lp		R	# ...but not those logs
=/tmp		N	# just the directory, not its files
.fi
.RE
.LP
Note the difference between pruning (via "!") and ignoring everything
(via "N" template):  ignoring everything in a directory still monitors
for added and deleted files, but pruning a directory will prevent Tripwire
from even looking in the specified directory for any changes.
.LP
.I Hint: 
Is Tripwire running too slowly?  Modify your \fItw.config\fP entries
to use only a few signatures (e.g., signatures 1 and 5) when this
computationally-exorbitant protection is not needed.  (See README and
design document for further details.)
.SH PREPROCESSOR
.LP
Tripwire incorporates a general purpose preprocessor that parses
the \fItw.config\fP file in one-pass.  Available preprocessing directives
include file inclusion, macro defines, conditionals based upon
hostname or macros, and on-the-fly macro substitution.  These directives
provide C-preprocessor and m4-like capabilities.
.LP
The Tripwire preprocessor was included to allow its scalable use at 
large sites, allowing system administrators to reuse \fItw.config\fP files
by either including component files or having multiple machines share a
common \fItw.config\fI file.
.RS
.TP 30
.I @@ifhost HOSTNAME	
includes text until matching \fI@@endif\fP if the machine
hostname matches the specified \fIHOSTNAME\fP.  Remember that you must use
the exact hostname that \fIuname(1)\fP or \fIhostname(1)\fP returns.  
This usually implies that you must use the fully qualified hostname
(e.g., mentor.cc.purdue.edu).
.TP
.I @@ifnhost HOSTNAME
includes text until matching \fI@@endif\fP if the machine
hostname does not match the specified \fIHOSTNAME\fP.
.TP
.I @@else 
provides if-else semantics to preprocessor.
.TP
.I @@define VAR STRING
defines variable \fIVAR\fP to \fISTRING\fP.  If the second argument
is not provided, then a null-string is assigned to \fIVAR\fP.
.TP
.I @@undef VAR
clears the definition associated with variable \fIVAR\fP.
.TP
.I @@ifdef VAR
includes text until the matching \fI@@endif\fP or \fI@@else\fP if the
variable \fIVAR\fP has been defined.
.TP
.I @@ifndef VAR
includes text until matching \fI@@endif\fP or \fI@@else\fP if the
variable \fIVAR\fP has not been defined.
.TP
.I @@endif
closes up \fI@@ifhost\fP, \fI@@ifnhost\fP, \fI@@ifdef\fP, and
\fI@@ifndef\fP.
.TP
.I @@include ``PATHNAME''
reads in the specified source file.  The double-quotes are optional.
.TP
.I @@VAR
substitutes the definition of \fIVAR\fP with the \fI@@define\fP'ed
value.
.TP
.I @@{VAR}
substitutes the definition of \fIVAR\fP with the \fI@@define\fP'ed
value.
.RE
.LP
.I Example:
A host-dependent inclusion can be specified many ways so \fItw.config\fP
files can be shared among multiple machines.  So, if the machine 
"mentor.cc.purdue.edu" is the only machine that has a certain file,
you could use:
.RS
.nf
@@ifhost mentor.cc.purdue.edu
/etc/tw.log.mentor	R
@@endif

@@define ARCHIVE	R
/etc/tw.log		@@ARCHIVE
.fi
.RE
.SH CAVEATS
.LP
Although Tripwire provides hooks for ten different
signature routines, using all ten would certainly be overkill in
almost any imaginable situation.  However, having up to ten
signature routines in your signature arsenal allows system
administrators considerable flexibility in finding the balance
between performance and security for their specific site.  This
is the reason for supplying CRC-16 and CRC-32, which are trivially
simple to spoof.  These routines are not secure, but they are
faster than the message-digesting routines.
.SH DATABASE VERSIONS
.LP
Tripwire v1.0 used database version 1.  Database version 2 changed the
base-64 alphabet so that ``0'' retained its traditional value.
.LP
Database
version 3 changed the base-64 encoding so that all the bits were packed,
reducing the size of 160-bit signatures from 30 characters to 27
characters.
Tripwire v1.1 used database version 3.  The program
\fItwconvert\fP is provided to convert from the older database formats
to version 3.
.LP
Tripwire v1.2 uses database version 4, supporting signatures for
symbolic links and more consistent handling of entry numbers.  (Note
that \fItwconvert\fP cannot convert older database versions to
database version 4.  These databases will have to be regenerated.)
.SH SEE ALSO
tripwire(8), twconvert(8)
.SH AUTHOR
.nf
Gene Kim
Purdue University
gkim@cs.purdue.edu

Eugene Spafford
Purdue University
spaf@cs.purdue.edu
.fi



uClinux-dist/release_notes
Release Notes for uClinux for Blackfin Version 2005R4

Version: 2005R4

Date: Dec. 30, 2005
Development Environments

Tool Chain: Toolchain Release 2005R4

u-boot: uboot_1.1.3 Release 2005R4

Host platform: Suse Linux 9.2 or above

Target board: STAMP Board

Note: However, other similar host platforms are also supported, but they are not well tested by now.
Deliverables

Source files

uClinux_dist_2005R4.tar.bz2

Linux ELF file

linux-bf533
linux-bf537
linux-bf561

Compressed Linux image

uImage-bf533
uImage-bf537
uImage-bf561

This document

release_notes_2005R4.pdf

Compressed archive of test results

test_results_bf533_2005R4.tar.gz
test_results_bf537_2005R4.tar.gz
test_results_bf561_2005R4.tar.gz

Summary of test results

test_results_summary_2005R4

Fixes in release 2005R4

See tracker at:

http://blackfin.uclinux.org/tracker/index.php?group_id=17&atid=141
Improvements

   1.
      Support BF561-EZKIT and add a tool to load code to run in core B.
   2.
      Support Bluetechnix CM Platform, which is also based on BF533.
   3.
      Enable module support in default configuration and generate module dependency file for modprobe.
   4.
      Upgrade to mainline source code
          *
            Upgrade uClibc to version 0.9.27.
          *
            Upgrade Linphone to version 1.1.0
          *
            Upgrade LTP to 20051205 release.
   5.
      Reorganize the scheduling and signal handling.
          *
            Defer scheduling and signal handling from different interrupt handlers to interrupt level 15.
          *
            Deal with the deferred exception handler and the defered interrupt handler in the same level of system call handler.
          *
            Run schedule and do signal from each interrupt handler above level 14.
          *
            Move softirq handler to interrupt level 14
   6.
      Prevent double fault exception.
          *
            Make sure that in an exception handler, we use a stack that is covered by a CPLB.
          *
            For exceptions other than CPLB misses, lower our priority to IRQ5, where we can use the normal kernel stack and use other regular kernel functions.
   7.
      Port new applications, libraries and tools
          *
            samba
          *
            strace
          *
            SQLite
          *
            lmbench3.
          *
            whetstone.
          *
            network audio player
          *
            ncurses
   8.
      Work around new hardware anomalys.
          *
            Bypass the first DCLPB fault exception, only handle the second one if it is from the same fault address.
          *
            Do not put SSYNC,CSYNC or IDLE in the second to last instruction of a hardware loop.
          *
            Do not use DMA_RUN bit to detect completion of a DMA transfer.
   9.
      Change paths from bfinnommu to blackfin, per request from Linux Kernel Mailing list.
  10.
      Enable oprofiler support in kernel and Port oprofiler tools
  11.
      Enable interrupt latency benchmark
  12.
      Add new device drivers
          *
            Enable IDE driver
          *
            Enable 802.11 WLAN driver
          *
            Enable PPI driver
          *
            Enable NAND flash driver
          *
            Port YAFFS driver
          *
            Add a common SPI channel driver. All high level drivers access SPI device by this driver.
  13.
      ADI1836 audio driver improvement
          *
            Reduce dummy buffer size from 256 bytes to 4 bytes
          *
            No interrupts generated in dummy DMA transfer
          *
            When audio driver is not active, DMA is also stopped.
          *
            Enable I2S mode for audio input and output. Modify mixer for I2S mode.
          *
            Enable multichannel output. Add a mixer function to change the output channels when play stereo.
          *
            Enable play multi independent audio streams concurrently.
          *
            Make audio DMA stable.
  14.
      Kernel debugging improvement
          *
            Enable hardware single stepping for application debugging.
          *
            Report precise hardware error information
          *
            debug kernel start. Install a mini-execption handler (replacing the U-Boot one), turn the UART on, and dump out the SEQSTAT register.
          *
            Enable hardware trace buffer in dump
  15.
      Improve power management
          *
            Implement the suspend function in bf533_pm_suspend.
          *
            SDRAM enters self-refresh mode before deepsleep to save more power.
          *
            Enable PM on BF537
  16.
      Enable PPPoE
  17.
      Enable DMA in SPI driver.
  18.
      Add board specific Kconfig Option for PF port/pin, which is used for flash and ethernet sharing.
  19.
      Enhance msh to support funtions in scripts.
  20.
      Transmite bfin atomic testandset operation in exception 4 into a new syscall, sys_bfin_spinlock.

Known Issues

A full list of known issues can be found at:

http://blackfin.uclinux.org/tracker/index.php?group_id=17&atid=141
No 	Issue Title
891	tcp stress testing echo service
900	busybox/msh,cd,tar and pipes
960	NFS unstable under heavy load
991	Infrequent audio crash
1012	After running dpmc to_deepsleep testcase, the kernel would panic after a while
1014	kernel built with audio configuration sometimes would fail to boot up and hangs

There are also some issues in the LTP test cases. They are recorded as bug 532, 561, 567, 732, 735, 742, 744, 745, 890 and 1010.
Build Procedure

1. Install Toolchain Release 2005R4

Go to http://blackfin.uclinux.org/frs/?group_id=18 for more information


uClinux-dist/user/tcsh/tw.comp.c
/* $Header: /cvsroot/uclinux533/uClinux-dist/user/tcsh/tw.comp.c,v 1.2 2005/07/25 06:42:15 magicyang Exp $ */
/*
 * tw.comp.c: File completion builtin
 */
/*-
 * Copyright (c) 1980, 1991 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
#include "sh.h"

RCSID("$Id: tw.comp.c,v 1.2 2005/07/25 06:42:15 magicyang Exp $")

#include "tw.h"
#include "ed.h"
#include "tc.h"

/* #define TDEBUG */
struct varent completions;

static int 	 	  tw_result	__P((Char *, Char **));
static Char		**tw_find	__P((Char *, struct varent *, int));
static Char 		 *tw_tok	__P((Char *));
static bool	 	  tw_pos	__P((Char *, int));
static void	  	  tw_pr		__P((Char **));
static int	  	  tw_match	__P((Char *, Char *));
static void	 	  tw_prlist	__P((struct varent *));
static Char  		 *tw_dollar	__P((Char *,Char **, int, Char *, 
					     int, const char *));

/* docomplete():
 *	Add or list completions in the completion list
 */
/*ARGSUSED*/
void
docomplete(v, t)
    Char **v;
    struct command *t;
{
    register struct varent *vp;
    register Char *p;

    USE(t);
    v++;
    p = *v++;
    if (p == 0)
	tw_prlist(&completions);
    else if (*v == 0) {
	vp = adrof1(strip(p), &completions);
	if (vp)
	    tw_pr(vp->vec), xputchar('\n');
    }
    else
	set1(strip(p), saveblk(v), &completions, VAR_READWRITE);
} /* end docomplete */


/* douncomplete():
 *	Remove completions from the completion list
 */
/*ARGSUSED*/
void
douncomplete(v, t)
    Char **v;
    struct command *t;
{
    USE(t);
    unset1(v, &completions);
} /* end douncomplete */


/* tw_prlist():
 *	Pretty print a list of variables
 */
static void
tw_prlist(p)
    struct varent *p;
{
    register struct varent *c;

    if (setintr)
#ifdef BSDSIGS
	(void) sigsetmask(sigblock((sigmask_t) 0) & ~sigmask(SIGINT));
#else				/* BSDSIGS */
	(void) sigrelse(SIGINT);
#endif				/* BSDSIGS */

    for (;;) {
	while (p->v_left)
	    p = p->v_left;
x:
	if (p->v_parent == 0)	/* is it the header? */
	    return;
	xprintf("%s\t", short2str(p->v_name));
	tw_pr(p->vec);
	xputchar('\n');
	if (p->v_right) {
	    p = p->v_right;
	    continue;
	}
	do {
	    c = p;
	    p = p->v_parent;
	} while (p->v_right == c);
	goto x;
    }
} /* end tw_prlist */


/* tw_pr():
 *	Pretty print a completion, adding single quotes around 
 *	a completion argument and collapsing multiple spaces to one.
 */
static void
tw_pr(cmp)
    Char **cmp;
{
    bool sp, osp;
    Char *ptr;

    for (; *cmp; cmp++) {
	xputchar('\'');
	for (osp = 0, ptr = *cmp; *ptr; ptr++) {
	    sp = Isspace(*ptr);
	    if (sp && osp)
		continue;
	    xputchar(*ptr);
	    osp = sp;
	}
	xputchar('\'');
	if (cmp[1])
	    xputchar(' ');
    }
} /* end tw_pr */


/* tw_find():
 *	Find the first matching completion. 
 *	For commands we only look at names that start with -
 */
static Char **
tw_find(nam, vp, cmd)
    Char   *nam;
    register struct varent *vp;
    int cmd;
{
    register Char **rv;

    for (vp = vp->v_left; vp; vp = vp->v_right) {
	if (vp->v_left && (rv = tw_find(nam, vp, cmd)) != NULL)
	    return rv;
	if (cmd) {
	    if (vp->v_name[0] != '-')
		continue;
	    if (Gmatch(nam, &vp->v_name[1]) && vp->vec != NULL)
		return vp->vec;
	}
	else
	    if (Gmatch(nam, vp->v_name) && vp->vec != NULL)
		return vp->vec;
    }
    return NULL;
} /* end tw_find */


/* tw_pos():
 *	Return true if the position is within the specified range
 */
static bool
tw_pos(ran, wno)
    Char *ran;
    int	  wno;
{
    Char *p;

    if (ran[0] == '*' && ran[1] == '\0')
	return 1;

    for (p = ran; *p && *p != '-'; p++)
	continue;

    if (*p == '\0')			/* range == <number> */
	return wno == getn(ran);
    
    if (ran == p)			/* range = - <number> */
	return wno <= getn(&ran[1]);
    *p++ = '\0';

    if (*p == '\0')			/* range = <number> - */
	return getn(ran) <= wno;
    else				/* range = <number> - <number> */
	return (getn(ran) <= wno) && (wno <= getn(p));
	   
} /* end tw_pos */


/* tw_tok():
 *	Return the next word from string, unquoteing it.
 */
static Char *
tw_tok(str)
    Char *str;
{
    static Char *bf = NULL;

    if (str != NULL)
	bf = str;
    
    /* skip leading spaces */
    for (; *bf && Isspace(*bf); bf++)
	continue;

    for (str = bf; *bf && !Isspace(*bf); bf++) {
	if (ismeta(*bf))
	    return INVPTR;
	*bf = *bf & ~QUOTE;
    }
    if (*bf != '\0')
	*bf++ = '\0';

    return *str ? str : NULL;
} /* end tw_tok */


/* tw_match():
 *	Match a string against the pattern given.
 *	and return the number of matched characters
 *	in a prefix of the string.
 */
static int
tw_match(str, pat)
    Char *str, *pat;
{
    Char *estr;
    int rv = Gnmatch(str, pat, &estr);
#ifdef TDEBUG
    xprintf("Gnmatch(%s, ", short2str(str));
    xprintf("%s, ", short2str(pat));
    xprintf("%s) = %d [%d]\n", short2str(estr), rv, estr - str);
#endif /* TDEBUG */
    return (int) (rv ? estr - str : -1);
}


/* tw_result():
 *	Return what the completion action should be depending on the
 *	string
 */
static int
tw_result(act, pat)
    Char *act, **pat;
{
    int looking;
    static Char* res = NULL;

    if (res != NULL)
	xfree((ptr_t) res), res = NULL;

    switch (act[0] & ~QUOTE) {
    case 'X':
	looking = TW_COMPLETION;
	break;
    case 'S':
	looking = TW_SIGNAL;
	break;
    case 'a':
	looking = TW_ALIAS;
	break;
    case 'b':
	looking = TW_BINDING;
	break;
    case 'c':
	looking = TW_COMMAND;
	break;
    case 'C':
	looking = TW_PATH | TW_COMMAND;
	break;
    case 'd':
	looking = TW_DIRECTORY;
	break;
    case 'D':
	looking = TW_PATH | TW_DIRECTORY;
	break;
    case 'e':
	looking = TW_ENVVAR;
	break;
    case 'f':
	looking = TW_FILE;
	break;
#ifdef COMPAT
    case 'p':
#endif /* COMPAT */
    case 'F':
	looking = TW_PATH | TW_FILE;
	break;
    case 'g':
	looking = TW_GRPNAME;
	break;
    case 'j':
	looking = TW_JOB;
	break;
    case 'l':
	looking = TW_LIMIT;
	break;
    case 'n':
	looking = TW_NONE;
	break;
    case 's':
	looking = TW_SHELLVAR;
	break;
    case 't':
	looking = TW_TEXT;
	break;
    case 'T':
	looking = TW_PATH | TW_TEXT;
	break;
    case 'v':
	looking = TW_VARIABLE;
	break;
    case 'u':
	looking = TW_USER;
	break;
    case 'x':
	looking = TW_EXPLAIN;
	break;

    case '$':
	*pat = res = Strsave(&act[1]);
	(void) strip(res);
	return(TW_VARLIST);

    case '(':
	*pat = res = Strsave(&act[1]);
	if ((act = Strchr(res, ')')) != NULL)
	    *act = '\0';
	(void) strip(res);
	return TW_WORDLIST;

    case '`':
	res = Strsave(act);
	if ((act = Strchr(&res[1], '`')) != NULL)
	    *++act = '\0';
	
	if (didfds == 0) {
	    /*
	     * Make sure that we have some file descriptors to
	     * play with, so that the processes have at least 0, 1, 2
	     * open
	     */
	    (void) dcopy(SHIN, 0);
	    (void) dcopy(SHOUT, 1);
	    (void) dcopy(SHDIAG, 2);
	}
	if ((act = globone(res, G_APPEND)) != NULL) {
	    xfree((ptr_t) res), res = NULL;
	    *pat = res = Strsave(act);
	    xfree((ptr_t) act);
	    return TW_WORDLIST;
	}
	return TW_ZERO;

    default:
	stderror(ERR_COMPCOM, short2str(act));
	return TW_ZERO;
    }

    switch (act[1] & ~QUOTE) {
    case '\0':
	return looking;

    case ':':
	*pat = res = Strsave(&act[2]);
	(void) strip(res);
	return looking;

    default:
	stderror(ERR_COMPCOM, short2str(act));
	return TW_ZERO;
    }
} /* end tw_result */
		

/* tw_dollar():
 *	Expand $<n> args in buffer
 */
static Char *
tw_dollar(str, wl, nwl, buffer, sep, msg)
    Char *str, **wl;
    int nwl;
    Char *buffer;
    int sep;
    const char *msg;
{
    Char *sp, *bp = buffer, *ebp = &buffer[MAXPATHLEN];

    for (sp = str; *sp && *sp != sep && bp < ebp;)
	if (sp[0] == '$' && sp[1] == ':' && Isdigit(sp[sp[2] == '-' ? 3 : 2])) {
	    int num, neg = 0;
	    sp += 2;
	    if (*sp == '-') {
		neg = 1;
		sp++;
	    }
	    for (num = *sp++ - '0'; Isdigit(*sp); num += 10 * num + *sp++ - '0')
		continue;
	    if (neg)
		num = nwl - num - 1;
	    if (num >= 0 && num < nwl) {
		Char *ptr;
		for (ptr = wl[num]; *ptr && bp < ebp - 1; *bp++ = *ptr++)
		    continue;
		
	    }
	}
	else
	    *bp++ = *sp++;

    *bp = '\0';

    if (*sp++ == sep)
	return sp;

    stderror(ERR_COMPMIS, sep, msg, short2str(str));
    return --sp;
} /* end tw_dollar */
		

/* tw_complete():
 *	Return the appropriate completion for the command
 *
 *	valid completion strings are:
 *	p/<range>/<completion>/[<suffix>/]	positional
 *	c/<pattern>/<completion>/[<suffix>/]	current word ignore pattern
 *	C/<pattern>/<completion>/[<suffix>/]	current word with pattern
 *	n/<pattern>/<completion>/[<suffix>/]	next word
 *	N/<pattern>/<completion>/[<suffix>/]	next-next word
 */
int
tw_complete(line, word, pat, looking, suf)
    Char *line, **word, **pat;
    int looking, *suf;
{
    Char buf[MAXPATHLEN + 1], **vec, *ptr; 
    Char *wl[MAXPATHLEN/6];
    static Char nomatch[2] = { (Char) ~0, 0x00 };
    int wordno, n;

    copyn(buf, line, MAXPATHLEN);

    /* find the command */
    if ((wl[0] = tw_tok(buf)) == NULL || wl[0] == INVPTR)
	return TW_ZERO;

    /*
     * look for hardwired command completions using a globbing
     * search and for arguments using a normal search.
     */
    if ((vec = tw_find(wl[0], &completions, (looking == TW_COMMAND))) == NULL)
	return looking;

    /* tokenize the line one more time :-( */
    for (wordno = 1; (wl[wordno] = tw_tok(NULL)) != NULL &&
		      wl[wordno] != INVPTR; wordno++)
	continue;

    if (wl[wordno] == INVPTR)		/* Found a meta character */
	return TW_ZERO;			/* de-activate completions */
#ifdef TDEBUG
    {
	int i;
	for (i = 0; i < wordno; i++)
	    xprintf("'%s' ", short2str(wl[i]));
	xprintf("\n");
    }
#endif /* TDEBUG */

    /* if the current word is empty move the last word to the next */
    if (**word == '\0') {
	wl[wordno] = *word;
	wordno++;
    }
    wl[wordno] = NULL;
	

#ifdef TDEBUG
    xprintf("\r\n");
    xprintf("  w#: %d\n", wordno);
    xprintf("line: %s\n", short2str(line));
    xprintf(" cmd: %s\n", short2str(wl[0]));
    xprintf("word: %s\n", short2str(*word));
    xprintf("last: %s\n", wordno - 2 >= 0 ? short2str(wl[wordno-2]) : "n/a");
    xprintf("this: %s\n", wordno - 1 >= 0 ? short2str(wl[wordno-1]) : "n/a");
#endif /* TDEBUG */
    
    for (;vec != NULL && (ptr = vec[0]) != NULL; vec++) {
	Char  ran[MAXPATHLEN+1],/* The pattern or range X/<range>/XXXX/ */
	      com[MAXPATHLEN+1],/* The completion X/XXXXX/<completion>/ */
	     *pos = NULL;	/* scratch pointer 			*/
	int   cmd, sep;		/* the command and separator characters */

	if (ptr[0] == '\0')
	    continue;

#ifdef TDEBUG
	xprintf("match %s\n", short2str(ptr));
#endif /* TDEBUG */

	switch (cmd = ptr[0]) {
	case 'N':
	    pos = (wordno - 3 < 0) ? nomatch : wl[wordno - 3];
	    break;
	case 'n':
	    pos = (wordno - 2 < 0) ? nomatch : wl[wordno - 2];
	    break;
	case 'c':
	case 'C':
	    pos = (wordno - 1 < 0) ? nomatch : wl[wordno - 1];
	    break;
	case 'p':
	    break;
	default:
	    stderror(ERR_COMPINV, CGETS(27, 1, "command"), cmd);
	    return TW_ZERO;
	}

	sep = ptr[1];
	if (!Ispunct(sep)) {
	    stderror(ERR_COMPINV, CGETS(27, 2, "separator"), sep);
	    return TW_ZERO;
	}

	ptr = tw_dollar(&ptr[2], wl, wordno, ran, sep,
			CGETS(27, 3, "pattern"));
	if (ran[0] == '\0')	/* check for empty pattern (disallowed) */
	{
	    stderror(ERR_COMPINC, cmd == 'p' ?  CGETS(27, 4, "range") :
		     CGETS(27, 3, "pattern"), "");
	    return TW_ZERO;
	}

	ptr = tw_dollar(ptr, wl, wordno, com, sep, CGETS(27, 5, "completion")); 

	if (*ptr != '\0') {
	    if (*ptr == sep)
		*suf = ~0;
	    else
		*suf = *ptr;
	}
	else
	    *suf = '\0';

#ifdef TDEBUG
	xprintf("command:    %c\nseparator:  %c\n", cmd, sep);
	xprintf("pattern:    %s\n", short2str(ran));
	xprintf("completion: %s\n", short2str(com));
	xprintf("suffix:     ");
        switch (*suf) {
	case 0:
	    xprintf("*auto suffix*\n");
	    break;
	case ~0:
	    xprintf("*no suffix*\n");
	    break;
	default:
	    xprintf("%c\n", *suf);
	    break;
	}
#endif /* TDEBUG */

	switch (cmd) {
	case 'p':			/* positional completion */
#ifdef TDEBUG
	    xprintf("p: tw_pos(%s, %d) = ", short2str(ran), wordno - 1);
	    xprintf("%d\n", tw_pos(ran, wordno - 1));
#endif /* TDEBUG */
	    if (!tw_pos(ran, wordno - 1))
		continue;
	    return tw_result(com, pat);

	case 'N':			/* match with the next-next word */
	case 'n':			/* match with the next word */
	case 'c':			/* match with the current word */
	case 'C':
#ifdef TDEBUG
	    xprintf("%c: ", cmd);
#endif /* TDEBUG */
	    if ((n = tw_match(pos, ran)) < 0)
		continue;
	    if (cmd == 'c')
		*word += n;
	    return tw_result(com, pat);

	default:
	    return TW_ZERO;	/* Cannot happen */
	}
    }
    *suf = '\0';
    return TW_ZERO;
} /* end tw_complete */
